Change kubectl create to use dynamic client

krousey opened the pull request:
https://github.com/kubernetes/kubernetes/issues/16764 https://github.com/kubernetes/kubernetes/issues/3955

This is a series of changes to allow kubectl create to use discovery-based REST mapping and dynamic clients.

cc @kubernetes/sig-api-machinery

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
kubectl will no longer do client-side defaulting on create and replace.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30250)

<!-- Reviewable:end -->

Commit:
Expose dynamic client's content config

Commit:
Eliminate redundant dynamic client type

This will allow people to override the default parameter codec and still
pass the resulting client with something that accepts *dynamic.Client.

Commit:
Properly decode lists into VersionedObjects

Commit:
Add generic "List" type for all versions

Commit:
Add discovery mapper and dynamic typer to kubectl

Commit:
kubectl/resource: Remove thirdparty special casing

And also remove some duplicate code while I'm at it.

Commit:
Make kubectl create use dynamic clients

smarterclayton commented on pkg/client/typed/discovery/restmapper.go:
```diff
@@ -80,6 +80,8 @@ func NewRESTMapper(groupResources []*APIGroupResources, versionInterfaces meta.V
 				versionMapper.Add(gv.WithKind(resource.Kind+"List"), scope)
 			}
+			// TODO why is this type not in discovery (at least for "v1")
+			versionMapper.Add(gv.WithKind("List"), meta.RESTScopeRoot)
```
It's not a server resource?  It's just a local object.  Version mapper shouldn't be blocking this - the server should never have to know about Lists

krousey replied:
Yea, I felt really wrong typing this in, but without it, `kubectl create -f hack/testdata/list.yaml` fails. I am certainly open to other suggestions.

smarterclayton commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -251,6 +245,18 @@ func (dynamicCodec) Encode(obj runtime.Object, w io.Writer) error {
 }
 
+// ContentConfig returns a restclient.ContentConfig for dynamic types.
+func ContentConfig() restclient.ContentConfig {
```
hard-coding api.Codecs into this method is fairly wrong - especially since you use nothing but JSON.  You should just be using your own CodecFactory here and only register json.

krousey replied:
This is just a move of what was already there. Would this be an easy fix, or something for another PR?

smarterclayton replied:
That's fine for another PR

smarterclayton commented on pkg/kubectl/cmd/create.go:
```diff
@@ -156,9 +157,23 @@ func RunCreate(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *C
+	var obj runtime.Object
+	var err error
+	if info.Dynamic {
+		dc, err := info.DynamicClient()
```
Why not reuse helper?  Not clear to me how this code is more correct?

krousey replied:
The idea is that helper is largely redundant to the dynamic client, and we should eventually replace the use of helper with dynamic client.

But you are right, this is premature and doesn't belong in this PR. Trying it piecemeal like this really complicates the code. Helper should work just fine with a proper restclient for dynamic types for now. I will work on that.

smarterclayton replied:
K - i'm happy to help dig in and fix anything that is broken - TPR ended up significantly complicating this code and I suspect all of your fixes are things that TPR should fall under.  I.e., unversioned objects, dynamic clients, codec hiding the details of the code.

smarterclayton commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -344,6 +348,40 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
+			}
+
+			mapper := discovery.NewRESTMapper(groupResources,
+				func(unversioned.GroupVersion) (*meta.VersionInterfaces, error) {
```
This seems like something that should be in a different package as a helper method.  Probably just meta

krousey replied:
How about runtime since it's closely tied to the unstructured object?

smarterclayton replied:
Since you depend on returning meta you can't put it in runtime (would be a circular reference).  So meta is as close as you can get to runtime.

Also, at some point in the future (probably when we implement the partial object return stuff) we'll want to move unstructured out of runtime and into its own package.

smarterclayton commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -1235,3 +1273,72 @@ func (f *Factory) NewBuilder(thirdPartyDiscovery bool) *resource.Builder {
 	return resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true))
 }
+
+// TODO find a better place for dynamicObjectTyper
```
"pkg/api/meta"

krousey replied:
I am hesitant to add it there as it would make meta depend on discovery client. I can move the type of the parameter(`discovery.APIGroupResources`) to the rest of the discovery types and then move this. Does that sound ok?

smarterclayton replied:
Actually, can you move this to `pkg/client/typed/discovery`?

smarterclayton commented on pkg/kubectl/resource/mapper.go:
```diff
@@ -54,19 +56,21 @@ func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
-		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
+
+	obj, versioned := versions.Last(), versions.First()
+	if !m.dynamic {
```
This just highlights how broken third party resource handling is.  This should *not* be necessary.

smarterclayton replied:
We have interfaces that already abstract third party.

smarterclayton replied:
I'm not sure yet whether you have to fix this for your PR, but a dynamic resource without versioning and a third party resource are functionally equivalent.  I think the interfaces we have today should be almost sufficient to hide all of this from resource builder, which is why I bring it up.  I would have expect this pr to use helper (or identify what needs to be changed), provide the appropriate codec and clients down, and hide all of these details (preferably removing this code from knowing about third party).

We already have a way in codec to signal that objects aren't versioned, if versioned.First == versioned.Last.  I think that's all you need here to collapse dynamic objects with 3rd party.  And DynamicObjects() should handle third party resources (no generic command needs to know the difference).

krousey replied:
So the `dynamic` flag is serving as a behavioral split between using dynamic types or not... for everything. When this is `true`, all objects are being deserialized into `*runtime.Unstructured` or `*runtime.UnstructuredList`. So for create, this whole branch of code doesn't execute. Create behaves the same for thirdparty as it does for pods. This split is here because other commands don't yet use the dynamic codec.

I think I can get rid of the dynamic split in most places by not trying to use the dynamic client. But if the caller want's to use dynamic decoding, I need to somehow prevent the third party from decoding it into it's functionally equivalent object.

Maybe the right answer is a codec that can serialize both unstructured and third party?

smarterclayton replied:
We have the code that lets a codec know a type is third party - I would
expect it to be possible to be a wrapper codec that says `if third party {
call third party codec code } return call dynamic code`

On Tue, Aug 9, 2016 at 1:54 PM, krousey <notifications@github.com> wrote:

> In pkg/kubectl/resource/mapper.go
> <https://github.com/kubernetes/kubernetes/pull/30250#discussion_r74109865>
> :
>
> > -	var obj runtime.Object
> > -	var versioned runtime.Object
> > -	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
> > -		return nil, err
> > -	} else if isThirdParty {
> > -		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
> > -		versioned = obj
> > -		gvk = gvkOut
> > -	} else {
> > -		obj, versioned = versions.Last(), versions.First()
> > -	}
> > -	if err != nil {
> > -		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
> > +
> > +	obj, versioned := versions.Last(), versions.First()
> > +	if !m.dynamic {
>
> So the dynamic flag is serving as a behavioral split between using
> dynamic types or not... for everything. When this is true, all objects
> are being deserialized into *runtime.Unstructured or
> *runtime.UnstructuredList. So for create, this whole branch of code
> doesn't execute. Create behaves the same for thirdparty as it does for
> pods. This split is here because other commands don't yet use the dynamic
> codec.
>
> I think I can get rid of the dynamic split in most places by not trying to
> use the dynamic client. But if the caller want's to use dynamic decoding, I
> need to somehow prevent the third party from decoding it into it's
> functionally equivalent object.
>
> Maybe the right answer is a codec that can serialize both unstructured and
> third party?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30250/files/8622d5ce64b4c381409b5fe86351e232b5e8e817#r74109865>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p1OZW9ZWfb9yhJwHxP7I-LNdp5p3ks5qeL6-gaJpZM4Jfj_0>
> .
>

smarterclayton replied:
The RESTMapping is the one who should be providing the codec - I would
expect the kubectl factory to know unambiguously whether to give a third
party codec or a dynamic client codec to return.

On Tue, Aug 9, 2016 at 2:17 PM, Clayton Coleman <ccoleman@redhat.com> wrote:

> We have the code that lets a codec know a type is third party - I would
> expect it to be possible to be a wrapper codec that says `if third party {
> call third party codec code } return call dynamic code`
>
> On Tue, Aug 9, 2016 at 1:54 PM, krousey <notifications@github.com> wrote:
>
>> In pkg/kubectl/resource/mapper.go
>> <https://github.com/kubernetes/kubernetes/pull/30250#discussion_r74109865>
>> :
>>
>> > -	var obj runtime.Object
>> > -	var versioned runtime.Object
>> > -	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
>> > -		return nil, err
>> > -	} else if isThirdParty {
>> > -		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
>> > -		versioned = obj
>> > -		gvk = gvkOut
>> > -	} else {
>> > -		obj, versioned = versions.Last(), versions.First()
>> > -	}
>> > -	if err != nil {
>> > -		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
>> > +
>> > +	obj, versioned := versions.Last(), versions.First()
>> > +	if !m.dynamic {
>>
>> So the dynamic flag is serving as a behavioral split between using
>> dynamic types or not... for everything. When this is true, all objects
>> are being deserialized into *runtime.Unstructured or
>> *runtime.UnstructuredList. So for create, this whole branch of code
>> doesn't execute. Create behaves the same for thirdparty as it does for
>> pods. This split is here because other commands don't yet use the dynamic
>> codec.
>>
>> I think I can get rid of the dynamic split in most places by not trying
>> to use the dynamic client. But if the caller want's to use dynamic
>> decoding, I need to somehow prevent the third party from decoding it into
>> it's functionally equivalent object.
>>
>> Maybe the right answer is a codec that can serialize both unstructured
>> and third party?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/30250/files/8622d5ce64b4c381409b5fe86351e232b5e8e817#r74109865>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p1OZW9ZWfb9yhJwHxP7I-LNdp5p3ks5qeL6-gaJpZM4Jfj_0>
>> .
>>
>
>

krousey replied:
The codecs provided by the factory come from `ClientForMapping` whose only input is the GVK. There is code in there to handle third party based on the GVK and use a third party codec instead.

My needs are slightly different. If the caller wants to use dynamic decoding, that is the only codec they should use, regardless of the GVK. But I also need to leave the orginal codecs intact for commands that can't use the dynamic objects yet. (like describe, delete, ....). So the choice to use dynamic objects here isn't based on the mapping, but rather if the caller doesn't need the versioned types.

smarterclayton replied:
That could be DynamicClientForMapping.  I don't know that `dynamic` is the
best word though going forward - you really want an unstructured client.
All third party objects are unstructured as well.   UnstructuredObject()
and UnstructuredClientForMapping() seem very reasonable to have on factory.

On Tue, Aug 9, 2016 at 2:44 PM, krousey <notifications@github.com> wrote:

> In pkg/kubectl/resource/mapper.go
> <https://github.com/kubernetes/kubernetes/pull/30250#discussion_r74120315>
> :
>
> > -	var obj runtime.Object
> > -	var versioned runtime.Object
> > -	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
> > -		return nil, err
> > -	} else if isThirdParty {
> > -		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
> > -		versioned = obj
> > -		gvk = gvkOut
> > -	} else {
> > -		obj, versioned = versions.Last(), versions.First()
> > -	}
> > -	if err != nil {
> > -		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
> > +
> > +	obj, versioned := versions.Last(), versions.First()
> > +	if !m.dynamic {
>
> The codecs provided by the factory come from ClientForMapping whose only
> input is the GVK. There is code in there to handle third party based on the
> GVK and use a third party codec instead.
>
> My needs are slightly different. If the caller wants to use dynamic
> decoding, that is the only codec they should use, regardless of the GVK.
> But I also need to leave the orginal codecs intact for commands that can't
> use the dynamic objects yet. (like describe, delete, ....). So the choice
> to use dynamic objects here isn't based on the mapping, but rather if the
> caller doesn't need the versioned types.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30250/files/8622d5ce64b4c381409b5fe86351e232b5e8e817#r74120315>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-QDr2h1iyIEBd9Nyqjaw31CJLgMks5qeMqdgaJpZM4Jfj_0>
> .
>

smarterclayton replied:
The boolean to Object(discoveryDynamicAPIs) is also kind of questionable -
as we have more unstructured objects, everything that is known to the CLI
via code injection should come from Object(), and everything else should be
come from UnstructuredObject().  In that model we wouldn't need a boolean
for discovery, because discovery is implicit for unstructured and explicit
for objects.

Also, I just noticed this, but Object() needs to return an error and it's
totally wrong that it doesn't.

On Tue, Aug 9, 2016 at 2:54 PM, Clayton Coleman <ccoleman@redhat.com> wrote:

> That could be DynamicClientForMapping.  I don't know that `dynamic` is the
> best word though going forward - you really want an unstructured client.
> All third party objects are unstructured as well.   UnstructuredObject()
> and UnstructuredClientForMapping() seem very reasonable to have on factory.
>
> On Tue, Aug 9, 2016 at 2:44 PM, krousey <notifications@github.com> wrote:
>
>> In pkg/kubectl/resource/mapper.go
>> <https://github.com/kubernetes/kubernetes/pull/30250#discussion_r74120315>
>> :
>>
>> > -	var obj runtime.Object
>> > -	var versioned runtime.Object
>> > -	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
>> > -		return nil, err
>> > -	} else if isThirdParty {
>> > -		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
>> > -		versioned = obj
>> > -		gvk = gvkOut
>> > -	} else {
>> > -		obj, versioned = versions.Last(), versions.First()
>> > -	}
>> > -	if err != nil {
>> > -		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
>> > +
>> > +	obj, versioned := versions.Last(), versions.First()
>> > +	if !m.dynamic {
>>
>> The codecs provided by the factory come from ClientForMapping whose only
>> input is the GVK. There is code in there to handle third party based on the
>> GVK and use a third party codec instead.
>>
>> My needs are slightly different. If the caller wants to use dynamic
>> decoding, that is the only codec they should use, regardless of the GVK.
>> But I also need to leave the orginal codecs intact for commands that can't
>> use the dynamic objects yet. (like describe, delete, ....). So the choice
>> to use dynamic objects here isn't based on the mapping, but rather if the
>> caller doesn't need the versioned types.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/30250/files/8622d5ce64b4c381409b5fe86351e232b5e8e817#r74120315>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p-QDr2h1iyIEBd9Nyqjaw31CJLgMks5qeMqdgaJpZM4Jfj_0>
>> .
>>
>
>

krousey replied:
Good news. I just found and fixed a bug in the thirdparty codec that let's me delete this nastiness here. This function will no longer explicitly know about either thirdparty or dynamic codecs. Still cleaning up other items though.

smarterclayton commented on pkg/kubectl/resource/selector.go:
```diff
@@ -76,12 +84,72 @@ func (r *Selector) Visit(fn VisitorFunc) error {
+}
+
+func (r *Selector) list() (runtime.Object, error) {
+	if !r.Dynamic {
```
This seems very wrong - Helper is *already* supposed to be generic - doing this is redundant.

smarterclayton commented on pkg/kubectl/resource/selector.go:
```diff
@@ -76,12 +84,72 @@ func (r *Selector) Visit(fn VisitorFunc) error {
+		return NewHelper(r.Client, r.Mapping).List(r.Namespace, r.ResourceMapping().GroupVersionKind.GroupVersion().String(), r.Selector, r.Export)
+	}
+
+	dc, err := r.dynamicClient()
```
This shouldn't be necessary - the Client you have should be perfectly capable of handling this abstraction.

smarterclayton commented on pkg/kubectl/resource/visitor.go:
```diff
@@ -106,9 +113,44 @@ func (i *Info) Visit(fn VisitorFunc) error {
 // Get retrieves the object from the Namespace and Name fields
 func (i *Info) Get() (err error) {
-	obj, err := NewHelper(i.Client, i.Mapping).Get(i.Namespace, i.Name, i.Export)
+	var obj runtime.Object
```
Same comment here - helper should be able to handle generic objects, and we need to fix helper, not add these ifs

smarterclayton commented on pkg/kubectl/resource/visitor.go:
```diff
@@ -106,9 +113,44 @@ func (i *Info) Visit(fn VisitorFunc) error {
+}
+
+// APIResource returns and APIResource for the Info object.
+func (i *Info) APIResource() *unversioned.APIResource {
```
Who would use this besides Info?

wojtek-t commented:
@smarterclayton @krousey - currently dynamic client is not using protobufs at all, so this may result in performance degradation. We should probably try to support protobufs in dynamic client too (based on some discovery?) WDYT?

deads2k commented on pkg/client/restclient/client.go:
```diff
@@ -114,6 +114,26 @@ func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ContentConf
+		return nil, err
+	}
+
+	newRC := *c
```
This shallow copy worries me.  We're still sharing the URL pointer, the client, BackoffManager func.

krousey replied:
You're right, I'll change this to call back into `NewRESTClient`

smarterclayton commented:
I think we can - part of my comments above would be correcting this code so that resource.Builder doesn't have to know about the details of all of that (it's hidden under Codec) which means that we can in the future use proto for those things.  However priority for client is not to use protobuf for the near term.

smarterclayton commented:
I can imagine interactions where you want to do reads using protobuf and then update an object via the dynamic client.  I just don't want to let the interfaces get mixed up - Codec hides those details from RESTClient, not the other way around.

Commit:
Fix thirdparty codec to return GVK

krousey commented on pkg/kubectl/resource/mapper.go:
```diff
@@ -54,20 +53,8 @@ func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
-		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
-	}
+
+	obj, versioned := versions.Last(), versions.First()
```
@smarterclayton does this look better?

smarterclayton replied:
Yes, a million times.

krousey commented:
@smarterclayton I've reworked this a good bit, and I think it's in a good spot for another pass.

krousey commented:
@smarterclayton Any chance you could look at this today? I'd really like to get it in before the code freeze.

smarterclayton commented:
A few more comments, this looks much cleaner and fits with the interfaces in place.  Having to add list to the discovery mapper seems wrong, can you open an issue for that and @ the api-machinery SIG on it so we can clean it up next release?  If I get the partial resources stuff and cross object version stuff nailed down that also has to get solved.

Do the third party resource tests adequately cover use of TPR in kubectl (to where you're confident the change in resource.Builder is safe?)

krousey commented:
@smarterclayton `resource.Builder` didn't change. I assume you're referring to `resource.Mapper` where I removed the third party special case. [test-cmd.sh:1017](https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-cmd.sh#L1017) registers 2 resources, creates, gets, and deletes them. I don't know how much more complicated we can get. Does that seem sufficient?

Also, now that you're happy with this (except for moves and follow up issues), mind if I change the update command in this PR too?

smarterclayton commented:
I think test-cmd is sufficient.  Sure - makes it easier to review.

krousey commented:
Fun fact: all of our objects should be `json.Marshal`-able. Otherwise `kubectl`'s annotation config saving is wrong. So you're going to see that in the next update to this PR.

smarterclayton commented:
Definitely internal objects should fail / be horribly incompatible if
someone tries to marshal them.  Did you find the apply internal version bug?

On Mon, Aug 15, 2016 at 6:27 PM, krousey <notifications@github.com> wrote:

> Fun fact: all of our objects should be json.Marshal-able. Otherwise
> kubectl's annotation config saving is wrong. So you're going to see that
> in the next update to this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30250#issuecomment-239948422>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p1exBpx7bE7SHD3IxtRGw-3IPLFvks5qgOfMgaJpZM4Jfj_0>
> .
>

krousey commented:
@smarterclayton I don't think so. My change is only using a single `runtime.Unstructured` object. So `info.Object == info.VersionedObject`. And the unstructured types do not marshal to the appropriate JSON by default. There's also a bunch of comments in the apply code warning of the assumption.

I'm just adding marshal handlers to the `Unstructured` and `UnstructuredList`.

Commit:
Add JSON encoding handlers to unstructured objects

Commit:
Make kubectl replace use dynamic clients

krousey commented:
@smarterclayton I think I addressed the remaining issues you had (the moves in to meta and discovery). I added some json marshaling to unstructured types (runtime/types.go) for proper annotations. I also migrated the replace command. Please take a look when you can.

Also... is this release note worthy?

krousey commented:
@k8s-bot test this

krousey commented:
@smarterclayton one last pass?

krousey commented:
@k8s-bot test this issue: IGNORE

smarterclayton commented:
I think this is release note worthy, since it changes user facing behavior of create.

smarterclayton commented:
Everything LGTM.  Any other items you felt were worthy of an additional look?

krousey commented:
@smarterclayton Nope, I think this is done. Thanks!
